Update DsUtils.cs

shell-with-it opened the pull request:
Updated getDomainSearchRoot to handle alternate UPN suffixes for domain. 

When Global Catalog is not used and the user upn contains a domain that is an alternate UPN suffix, SID lookup fails.

shell-with-it committed:
Update DsUtils.cs

Updated getDomainSearchRoot to handle alternate UPNs suffixes for domain.

Crypt32 commented on src/ADCS.SidExtension.PolicyModule/DsUtils.cs:
```diff
@@ -57,13 +57,24 @@ static class DsUtils {
             // upn
-            Int32 splitIndex = name.IndexOf("@", StringComparison.Ordinal);
-            domainTokens = name.Substring(splitIndex + 1).Split('.');
+            String domainPart = name.Substring(splitIndex + 1);
```
`splitIndex` is not defined here. I believe that you accidentally removed a line where this variable is defined and initialized.

shell-with-it replied:
You're correct, your original definition should've remained. I have also updated it to iterate through each domain in the forest. I'm fairly new to C# so please forgive me.

Crypt32 commented on src/ADCS.SidExtension.PolicyModule/DsUtils.cs:
```diff
@@ -57,13 +57,24 @@ static class DsUtils {
-            domainTokens = name.Substring(splitIndex + 1).Split('.');
+            String domainPart = name.Substring(splitIndex + 1);
+            using (var forest = Forest.GetCurrentForest()) {
+                using (var domain = forest.Domains.FirstOrDefault(d => d.Forest.Name.Equals(domainPart, StringComparison.OrdinalIgnoreCase))) {
```
I'm not sure how it is supposed to work. All domains share same forest name, so `forest.Domain.FirstOrDefault` call makes little to no sense, because first domain entry will match the predicate and others are not evaluated. And there always will be one object -- forest root domain. In multi-domain forest it is going to fail. IIRC, domains in `forest.Domains` collection are ordered by name, so the predicate will match the first alphabetically ordered domain in the list. And it doesn't necessary have to be a domain where this particular UPN lives in.

In other words, I understand the problem, however this PR addresses the issue in a single-domain forest. Once you get two domains, it is going to fail.

In case of custom UPN suffix, it may be necessary to query all domains in a forest until match is found.

shell-with-it committed:
Update DsUtils.cs

Updated to search all domains in forest instead of pulling first from forest, re-added splitIndex definition.

shell-with-it committed:
Update DsUtils.cs

Added ; after null

shell-with-it reviewed (commented):
Added ;

Crypt32 commented on src/ADCS.SidExtension.PolicyModule/DsUtils.cs:
```diff
@@ -58,12 +58,23 @@ static class DsUtils {
+            {
+                foreach (Domain domain in forest.Domains)
+                {
+                    if (domain.GetDirectoryEntry().Properties["uPNSuffixes"].Contains(domainPart)) {
```
I don't think that `uPNSuffixes` attribute is part of domain object. In fact, when you create a new UPN suffix, it is created for entire forest and stored in configuration naming context which isn't bound to any particular domain it is bound to forest root domain where configuration naming context is stored. In addition, this information can be cached in order to boost the performance. This will require additional code to support this. Let me think about this and I will address this issue. Anyways, thanks for raising this point and your efforts!

shell-with-it replied:
Thank you for the Policy module! Glad I could get this on your radar.
